### PR TITLE
WIP

### PR DESCRIPTION
**Pull request by isuruf**
Here's my attempt to create the classes

### CODE REVIEW

File: src/types.jl [hunk 1]
```diff
@@ -52,27 +52,44 @@ convert(::Type{Basic}, x::Rational) = Basic(BasicInteger(num(x)) / BasicInteger(
...
+for (k,v) in basic_types
+    tname = symbol("Basic$k")
+    pname = symbol("$v")
     @eval begin
-        type $tname <: BasicType
+        type $tname <: $pname
             x::Basic
         end
-        _basic_types[$k] = $tname
+        t = ccall((:basic_get_class_id, :libsymengine), UInt, (Ptr{Int8},), "$k")
```
Comment (isuruf) [inline src/types.jl]: This line gives an error

**Comment by isuruf** [comment]
This is working now. @jverzani, can you take a look?

**Comment by jverzani** [comment]
I'm not sure that is the way to go with this. Since I changed the types to be parameterized, it might be better to make type unions:

```
BasicTrigType = Union{SymEngine.BasicType{Val{:Cos}},SymEngine.BasicType{Val{:Sin}}}}
```

(Using some trick to cut down on the verbosity.) This would be no harder to maintain, and gives the user the ability to fine tune their methods should they want. I'll push a change my WIP to see how it would look.

**Comment by jverzani** [comment]
Take a look here to see what i mean: https://github.com/jverzani/SymEngine.jl/blob/pull-request/8c92fb77/src/types.jl#L95

The fine-grained types BasicType{Val{:Integer}} etc are created as needed from the symengine object, and these unions make it easy to control dispatch.

**Pull request closed without merging**